labs/b9.md

```diff
@@ -0,0 +1,178 @@
+
+For further reading, I recommend just reading the [official documentation](https://docs.docker.com/) so you can see what is possible with the Docker container format.    
+
+#### Puppet
+
+To get some hands on experience with Puppet, we are going to write a very basic manifest that pulls quotes from the internet and appends them to a file every two minutes. Although you were taught that in production environments we have the puppet master running on its own server, for simplicity you will apply the manifest locally, making you both the puppet master and the puppet agent. If you are confused about the Puppet vocabulary, check out slide 21.
+
+First things first. We plan to pull quotes from the web, so we need to ensure the `wget` package is installed on our system. Next we need to create a user, 'quotes', that runs this command for us. Take note of the dependency between the 'quotes' user and the 'quotegather' group existing. 
+
+Because we have full control over our user, let's make its home directory `/tmp`. Since this user has one purpose, to pull quotes from the web, they don't need a login shell. Go ahead and set it to `/bin/false`.
+
+Now we need a `cron` resource that grabs the quotes for us. I went ahead and filled in the command, so all you need to do is specify which user is to run the cron job and the interval at which it runs (every 2 minutes).
+
+Below I've provided a skeleton manifest `quotes.pp`. This can also be found in `~/decal-labs/b9/puppet/quotes.pp`, but comments are included here for your convenience.
+
+```puppet
+package { '': } # What package is required for this script to run?
+
+group { 'quotegather':
+  ensure  => '', # Do we want the group on this system?
+}
+
+user { 'quotes':
+  ensure  => '', # Do we want the user on this system?
+  gid     => , # Note that we can pass a group name to gid
+  home    => '', 
+  shell   => '',
+  # Do not create the user quotes unless the quotegather group exists
+  require => Group['quotegather'], 
+}
+
+cron { 'getquote':
+  user     => '',
+  # You can ignore this, it pulls the quotes from an API and appends
+  # them to a file.
+  command  => "curl -X POST http://api.forismatic.com/api/1.0/ -d 'method=getQuote' -d 'format=text' -d 'lang=en' >> quotes && echo >> quotes",
+  minute   => '', # Hint: passing * means this runs every minute
+}
+```
+Once you have completed your manifest, you can apply the changes to your system with the command `sudo puppet apply quotes.pp`. 
+
+Some debugging tips:
+- Read the [Offical Puppet Documentation](https://puppet.com/docs/puppet/5.5/type.html) about different resources if you are confused about commands
```

Comment:
I'd link to the 4.8 docs, since that's the version they have installed. (Shouldn't really be any differences, but seems more accurate than using 5.5)